### PR TITLE
Simulate uplink API

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

References #2234 

#### Changes

- Add `SimulateUplink` to `AppAs`
- Add support for simulated traffic to the AS

#### Testing

I've created an app `app1` and a device `dev11` (bad keyboard) and then simulated the uplink as follows:
```bash
curl --location --request POST 'http://localhost:1885/api/v3/as/applications/app1/devices/dev11/up/simulate' \
--header 'Authorization: Bearer $KEY' \
--header 'Content-Type: application/json' \
--data-raw '{
  "@type": "type.googleapis.com/ttn.lorawan.v3.ApplicationUp",
  "end_device_ids": {
    "device_id": "dev111",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "0000000000000001",
    "join_eui": "0000000000000000",
    "dev_addr": "270000DD"
  },
  "correlation_ids": [
    "gs:conn:01ED482WRPY2BABY4TYZV57RJG",
    "gs:uplink:01ED9B93M49J8P3FQXCGGCYTGX",
    "ns:uplink:01ED9B93MH00CTY41A6KF674E4",
    "rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01ED9B93MH293RE2TR3F1WMHEG"
  ],
  "uplink_message": {
    "session_key_id": "BXGsg614fdmYH7efd+fRvB==",
    "f_port": 17,
    "f_cnt": 23787,
    "frm_payload": "AhJF8HTI3khf",
    "received_at": "2020-07-15T13:40:31.505981496Z"
  }
}'
```

In MQTT I've then received the following uplink (notice the `decoded_payload`):
```json
{
  "end_device_ids" : {
    "device_id" : "dev11",
    "application_ids" : {
      "application_id" : "app1"
    },
    "dev_eui" : "0000000000000001",
    "join_eui" : "0000000000000000",
    "dev_addr" : "270000DD"
  },
  "correlation_ids" : [ "as:up:01ED9J8BM3G5XK8JRZJJ66FN4W", "gs:conn:01ED482WRPY2BABY4TYZV57RJG", "gs:uplink:01ED9B93M49J8P3FQXCGGCYTGX", "ns:uplink:01ED9B93MH00CTY41A6KF674E4", "rpc:/ttn.lorawan.v3.AppAs/SimulateUplink:01ED9J8BK1MV0TF76HEQ56EZS5", "rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01ED9B93MH293RE2TR3F1WMHEG" ],
  "received_at" : "2020-07-15T15:42:26.947920950Z",
  "uplink_message" : {
    "session_key_id" : "BXGsg614fdmYH7efd+fRvA==",
    "f_port" : 17,
    "f_cnt" : 23787,
    "frm_payload" : "AhJF8HTI3khf",
    "decoded_payload" : {
      "f_port" : 17,
      "raw" : [ 2, 18, 69, 240, 116, 200, 222, 72, 95 ]
    },
    "settings" : {
      "data_rate" : { }
    },
    "received_at" : "2020-07-15T13:40:31.505981496Z"
  }
}
```

##### Regressions

We've introduced a branch in `handleUp` in order to check if the traffic is simulated or not. I consider it not to affect the hot path performance, but it's something to keep in mind. We can avoid walking the whole context up to the root by using `WithSimulatedTraffic(ctx, false)` in the NS link.

#### Notes for Reviewers

As you can see in the test/example, you may leave out any field (like `settings` or `rx_metadata`). You can also submit other message types (`JoinAccept`, `DownlinkNack`, etc.).

You can test this by replaying traffic from the Console events tab too. Just copy paste the contents of the event into the body and you are ready to go.

@kschiffer please let me know if you want the AS to generate fake metadata for you. I don't think it's required tho.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -31,7 +31,7 @@ require (
 	github.com/chrj/smtpd v0.1.2
 	github.com/disintegration/imaging v1.6.2
 	github.com/eclipse/paho.mqtt.golang v1.2.0
-	github.com/envoyproxy/protoc-gen-validate v0.3.0-java
+	github.com/envoyproxy/protoc-gen-validate v0.1.0
```
Comment (rvolosatovs) [inline go.mod:34]: Revert

File: tools/go.sum [hunk 2]
```diff
@@ -251,6 +251,7 @@ github.com/eknkc/amber v0.0.0-20171010120322-cdade1c07385/go.mod h1:0vRUJqYpeSZi
 github.com/envoyproxy/go-control-plane v0.9.0/go.mod h1:YTl/9mNaCwkRvm6d1a2C3ymFceY/DCBVvsKhRF0iEA4=
 github.com/envoyproxy/go-control-plane v0.9.1-0.20191026205805-5f8ba28d4473/go.mod h1:YTl/9mNaCwkRvm6d1a2C3ymFceY/DCBVvsKhRF0iEA4=
 github.com/envoyproxy/go-control-plane v0.9.4/go.mod h1:6rpuAdCZL397s3pYoYcLgu1mIlRU8Am5FuJP05cCM98=
+github.com/envoyproxy/protoc-gen-validate v0.1.0 h1:EQciDnbrYxy13PgWoY8AqoxGiPrpgBZ1R8UNe3ddc+A=
```
Comment (rvolosatovs) [inline tools/go.sum:254]: IDEM

File: api/applicationserver.proto [hunk 3]
```diff
@@ -131,6 +131,12 @@ service AppAs {
       get: "/as/applications/{application_id}/mqtt-connection-info"
     };
   };
+  rpc SimulateUplink(ApplicationUp) returns (google.protobuf.Empty) {
+    option (google.api.http) = {
+      post: "/as/applications/{end_device_ids.application_ids.application_id}/devices/{end_device_ids.device_id}/up/push",
```
Comment (johanstokking) [inline api/applicationserver.proto:136]: I would make this `/simulate`

File: pkg/applicationserver/applicationserver.go [hunk 4]
```diff
@@ -540,6 +543,15 @@ func (as *ApplicationServer) handleUp(ctx context.Context, up *ttnpb.Application
 	}
 }
 
+func (as *ApplicationServer) handleSimulatedUp(ctx context.Context, up *ttnpb.ApplicationUp, link *link) error {
+	switch p := up.Up.(type) {
+	case *ttnpb.ApplicationUp_UplinkMessage:
```
Comment (johanstokking) [inline pkg/applicationserver/applicationserver.go:548]: So can we or can we not simulate other upstream message types like in the original comment?

File: pkg/applicationserver/io/grpc/grpc.go [hunk 5]
```diff
@@ -149,3 +149,14 @@ func (s *impl) GetMQTTConnectionInfo(ctx context.Context, ids *ttnpb.Application
 		Username:         unique.ID(ctx, *ids),
 	}, nil
 }
+
+func (s *impl) SimulateUplink(ctx context.Context, up *ttnpb.ApplicationUp) (*pbtypes.Empty, error) {
+	if err := rights.RequireApplication(ctx, up.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_TRAFFIC_UP_WRITE); err != nil {
```
Comment (johanstokking) [inline pkg/applicationserver/io/grpc/grpc.go:154]: The right is ok; no need for separate simulation right

File: pkg/applicationserver/applicationserver.go (hunk 4 above)
Comment (adriansmares) [reply pkg/applicationserver/applicationserver.go:548]: We can. `handleSimulatedUp` does to simulated traffic what `handleUp` does to real traffic - process the contents of the `ApplicationUp`. For real traffic this means decrypt and decode for `UplinkMessage`, recalculating the queue on `DownlinkQueueInvalidated` and so forth. The simulated traffic does not need any special processing when it comes to other message types, so there are no switch cases for these.

Comment (johanstokking) [reply pkg/applicationserver/applicationserver.go:551]: OK yeah I see now, thanks

File: pkg/applicationserver/io/grpc/grpc.go [hunk 6]
```diff
@@ -149,3 +149,14 @@ func (s *impl) GetMQTTConnectionInfo(ctx context.Context, ids *ttnpb.Application
...
+		return nil, err
+	}
+	ctx = io.WithSimulatedTraffic(ctx, true)
+	if err := s.server.SendUp(ctx, up); err != nil {
+		return nil, err
+	}
```
Comment (johanstokking) [inline pkg/applicationserver/io/grpc/grpc.go:160]: You could simplify it now and set `up.Simulated = true` here and remove the signaling via context.

**Pull request merged**